language-plutus-core/plutus-ir/Language/PlutusIR/Transform/Inline.hs

```diff
@@ -0,0 +1,109 @@
+inline!
+
+There are various ways to fix this, but ultimately they all involve doing
+substitutions multiple times (and hence multiple tree traversals). We'd ideally
+like to minimize this.
+
+The approach we take is to:
+- Collect all the substitutions we want to do in the whole program (needs
+global uniqueness).
+- Apply all the substitutions bottom up, at each point applying the substitution
+to the substituted term until we can't do any more.
+
+This is not terribly efficient, since we keep trying to re-apply our substitutions,
+and as we go up the tree we will keep doing this. So we could probably do this better
+but we don't try at the moment.
+-}
+
+{- Note [Inlining criteria]
+What gets inlined? We don't really care about performance here, so we're really just
+angling to simplify the code without making things worse.
+
+The obvious candidates are tiny things like builtins, variables, or constants.
+We could also consider inlining variables with arbitrary RHSs that are used only
+once, but we don't do that currently.
+-}
+
+{- Note [Inlining bindings]
+We can inline term and type bindings, we can't do anything with datatype bindings.
+
+We don't actually inline type bindings at the moment, mostly because I think it
+won't get us much as they aren't created very often..
+-}
+
+type Inlinings tyname name uni a = UniqueMap TermUnique (Term tyname name uni a)
+
+-- | Inline simple bindings. Relies on global uniqueness, and preserves it.
+inline
+    :: (HasUnique name TermUnique)
+    => Term tyname name uni a
+    -> Term tyname name uni a
+inline t =
+    let inlinings = collectInlinings t
+    -- Do the substitutions, iterating until we can't do any more.
+    -- See Note [Repeated inlining]
+    in termSubstNamesRepeat (\x -> lookupName x inlinings) t
+
+-- | Is this a an utterly trivial term which might as well be inlined?
+trivialTerm :: Term tyname name uni a -> Bool
+trivialTerm = \case
+    Builtin{} -> True
+    Var{} -> True
+    -- TODO: Should this depend on the size of the constant?
+    Constant{} -> True
+    _ -> False
+
+collectBindingInlinings
+    :: (HasUnique name TermUnique)
+    => Recursivity
+    -> Binding tyname name uni a
+    -> Inlinings tyname name uni a
+collectBindingInlinings Rec = const  mempty
+collectBindingInlinings NonRec = \case
+    -- See Note [Inlining criteria]
+    -- See Note [Inlining bindings]
+    TermBind _ Strict (VarDecl _ n _) rhs | trivialTerm rhs -> insertByName n rhs mempty
+    _ -> mempty
+
+collectInlinings
+    :: (HasUnique name TermUnique)
+    => Term tyname name uni a
+    -> Inlinings tyname name uni a
+collectInlinings x =
+    let subtermInlinings = foldMapOf termSubterms collectInlinings x
+        bindingInlinings = case x of
+            -- Can't quite just compose the termBindings traversal, since
+            -- we need to know the recursivity
+            Let _ r bs _ -> foldMap (collectBindingInlinings r) bs
```

Comment:
Why not match on the recursivity here, so that if it's `Rec`, then you don't need to process `bs` at all?
What does wrong if recursive lets are processed the same way as non-recursive ones?

Reply:
Given the way we do substitutions, I wanted to avoid non-terminating inlining where we unroll recursive definitions forever. That would be a little tricky to trigger currently, but generally recursive bindings seem risky.

Reply:
As for the matching, I don't think it makes much difference either way...

Reply:
I guess subconsciously I was thinking that *if* we did handle recursive bindings, we might have to do something special in the binding collection case.

Reply:
> I guess subconsciously I was thinking that if we did handle recursive bindings, we might have to do something special in the binding collection case.

I'd still prefer two different functions then (one handling the recursive case, the other handling the non-recursive one). The current solution reads as "whatever your recursivity is, you'll need to `foldMap` over the list of bindings", which kinda states more about the recursive case than we actually need/know.

Doesn't matter much, I agree.